File: LeetCode problem set/20_Valid_Parentheses.cpp
// https://leetcode.com/problems/valid-parentheses/
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    bool isValid(string s) {
      stack<char> st;
      for (int i = 0; i < s.size(); i++) {
        char now = s[i];
        if(now == '(' || now == '{' || now == '[') {
          st.push(s[i]);
        } else {
          if(st.empty()) return false;
          if( now == ')' && st.top()=='(' )
            st.pop();
         else if( now == '}' && st.top()=='{' )
            st.pop();
         else if( now == ']' && st.top()=='[' )
            st.pop();
          else 
            return false;
        }
      }
      if(st.empty()) return true;
      else return false;
    }
};

int main() {
  ios_base::sync_with_stdio(0), cin.tie(0), cout.tie(0);
  Solution sol;
  cout<<sol.isValid("()[]{}")<<endl;
  return 0;
}

File: LeetCode problem set/20_Valid_Parentheses.js
// https://leetcode.com/problems/valid-parentheses/

const isValid = (s) => {
  const stack = [];
  const map = {
    "(": ")",
    "{": "}",
    "[": "]",
  };

  for (let i = 0; i < s.length; i++) {
    let char = s[i];
    if (map[char]) {
      stack.push(char);
    } else {
      let sz = stack.length;
      if (sz === 0) return false;
      let last = stack[sz - 1];
      if (last === "(" && char === ")") stack.pop();
      else if (last === "{" && char === "}") stack.pop();
      else if (last === "[" && char === "]") stack.pop();
      else return false;
    }
  }
  if (stack.length === 0) return true;
  else return false;
};

console.log(isValid("(){}[]"));
